natlas-server/app/__init__.py

```diff
@@ -39,13 +39,78 @@ def unauthorized():
 		return redirect(url_for('main.index'))
 
 
-def create_app(config_class=Config, load_config=False):
+def load_natlas_config(app):
+	if not db.engine.has_table("config_item"):
+		return
+
+	updated_items = 0
+	from app.models import ConfigItem
+	# Look to see if any new config items were added that aren't currently in db
+	default_configs = config.get_defaults()
+	for key, item in default_configs:
+		conf_item = ConfigItem.query.filter_by(name=key).first()
+		if not conf_item:
+			conf_item = ConfigItem(name=key, type=item['type'], value=item['default'])
+			db.session.add(conf_item)
+			updated_items += 1
+		app.config[conf_item.name] = config.casted_value(conf_item.type, conf_item.value)
+	if updated_items > 0:
+		db.session.commit()
+
+
+def load_natlas_services(app):
+	if not db.engine.has_table("natlas_services"):
+		return
+	from app.models import NatlasServices
+	current_services = NatlasServices.query.order_by(NatlasServices.id.desc()).first()
+	if not current_services:
+		# Let's populate server defaults
+		defaultServices = open(os.path.join(app.config["BASEDIR"], "defaults/natlas-services")).read().rstrip('\r\n')
+		defaultSha = hashlib.sha256(defaultServices.encode()).hexdigest()
+		current_services = NatlasServices(sha256=defaultSha, services=defaultServices) # default values until we load something
```

Comment:
The sha should be computed by the NatlasServices class itself. What happens if you change the algorithm? You'd have to update in multiple places. This logic is not a concern of load_natlas_services, it's a concern of NatlasServices itself.

Reply:
I've moved this logic into the initialization of NatlasServices, but the column name in the database is still sha256. I'd rather not batch that change into the migration that's already part of this PR, but if we move to a multitenant model then we'll probably want to add a relationship to this table to link a given version of the services to a particular tenant, at which point would be a good time to also change the name from sha256 to hash.

Reply:
I think it's reasonable to defer supporting different hash algorithms for when we need it. My main goal was to make sure that we moved this code into the model. This change improves code quality, whereas supporting hash algorithms should be done when we specifically need that capability.